### ISSUE TITLE
Unexpected results w/ LayerNorm -- suspecting possible memory issue?

### ISSUE THREAD

**Issue by Apex95** [pytorch/pytorch#149301]
### 🐛 Describe the bug

I'm noticing an interesting behaviour of LayerNorm when applied to large 4d tensors (bf16) when normalized shape is an int (i.e., normalizing over the final dimension of the tensor).

What I'm seeing is that the size of the first dimension (batch size) can impact the normalized values of other samples. To be specific, if A and B are 2 input tensors (4d), where B = A[:-1], then after passing the LayerNorm layer, there's a difference between A[:-1] and B even though B is a subset of A. It's almost as if LayerNorm has some memory-access issue?

This does not happen if I run smaller tensors through this operation or if I run this through a manual normalization (using torch.mean() and unbiased torch.var()).

A code that reproduces this on A100/40GB would be something like:
```python
import torch
import torch.nn.functional as F

torch.manual_seed(1337)
device = torch.device('cuda')

USE_MANUAL_LAYERNORM = False

ln1 = torch.nn.LayerNorm(768).to(device).to(torch.bfloat16)
ln1.eval()

@torch.inference_mode()
def test():

    x = torch.rand((24, 493, 768), device=device, dtype=torch.bfloat16)
    x1 = x[:-1]
    print('-----------------------')
    print(f'x shape: {x.shape} | x1 shape: {x1.shape}')
    print('> Max Diff at input:\n', (x[:-1]-x1[:]).abs().max())

    x = torch.tanh(x)
    x1 = torch.tanh(x1)

    x = x[:, :, None, :] + x[:, None, :, :]
    x1 = x1[:, :, None, :] + x1[:, None, :, :]
    print('> Max Diff after broadcast:\n', (x[:-1]-x1[:]).abs().max())

    x = F.gelu(x)
    x1 = F.gelu(x1)
    print('> Max Diff after non-linearity:\n', (x[:-1]-x1[:]).abs().max())

    _x = ln1(x[:, :, 0, :])
    _x1 = ln1(x1[:, :, 0, :])
    print('> Max Diff after 3d layernorm:\n', (_x[:-1]-_x1[:]).abs().max())

    if USE_MANUAL_LAYERNORM:
        x = (x - x.mean(dim=-1, keepdim=True)) / torch.sqrt(torch.var(x, dim=-1, keepdim=True, correction=1)+1e-8)
        x1 = (x1 - x1.mean(dim=-1, keepdim=True)) / torch.sqrt(torch.var(x1, dim=-1, keepdim=True, correction=1)+1e-8)

        print(x[0,:2, :2, :2])
        print(x1[0,:2, :2, :2])
        print('> Max Diff after manual 4d layernorm:\n', (x[:-1]-x1[:]).abs().max())
    else:
        x = ln1(x)
        x1 = ln1(x1)

        print(x[0,:2, :2, :2])
        print(x1[0,:2, :2, :2])
        print('> Max Diff after 4d layernorm:\n', (x[:-1]-x1[:]).abs().max())

test()
```

Which yields, for a large tensor:
```x shape: torch.Size([24, 493, 768]) | x1 shape: torch.Size([23, 493, 768])
> Max Diff at input:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after broadcast:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after non-linearity:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after 3d layernorm:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
[...]
> Max Diff after 4d layernorm:
 tensor(0.6172, device='cuda:0', dtype=torch.bfloat16)
```

Doing this with `USE_MANUAL_LAYERNORM = True` gives:
```
x shape: torch.Size([24, 493, 768]) | x1 shape: torch.Size([23, 493, 768])
> Max Diff after backbone:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after broadcast:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after non-linearity:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after 3d layernorm:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
[...]
> Max Diff after manual 4d layernorm:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
```

Also, for a smaller tensor (i.e., x.shape = (24, 200, 768)):
```
x shape: torch.Size([24, 200, 768]) | x1 shape: torch.Size([23, 200, 768])
> Max Diff at input:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after broadcast:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after non-linearity:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
> Max Diff after 3d layernorm:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
[...]
> Max Diff after 4d layernorm:
 tensor(0., device='cuda:0', dtype=torch.bfloat16)
```

Please let me know if there's any mistake in my understanding of this.

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.31

Python version: 3.11.11 | packaged by conda-forge | (main, Mar  3 2025, 20:43:55) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.10.0-34-cloud-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        46 bits physical, 48 bits virtual
CPU(s):                               12
On-line CPU(s) list:                  0-11
Thread(s) per core:                   2
Core(s) per socket:                   6
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                85
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                             7
CPU MHz:                              2200.240
BogoMIPS:                             4400.48
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            192 KiB
L1i cache:                            192 KiB
L2 cache:                             6 MiB
L3 cache:                             38.5 MiB
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] triton==3.2.0
[conda] numpy                     2.2.3                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi
```

**Comment by albanD** [comment]
Could this be due to https://pytorch.org/docs/stable/notes/numerical_accuracy.html ?

**Comment by Apex95** [comment]
I've considered this but I don't understand why the results would be different when using the same LayerNorm on the same input. The example above shows that adding a new sample in the batch (going from batch size 23 (x1) to batch size 24 (x)) changes the result of the normalization for common inputs as well. My assumption was that LayerNorm is batch independent and thus `LayerNorm(x)[:-1] == LayerNorm(x[:-1])`

In this case I'm getting slightly different layernorm values for, e.g., x[0] and x1[0], even if they're the same and are processed in the same way.

**Comment by Apex95** [comment]
I've also attached here a memcheck for the batch size 26 run which kind of shows a couple of memory leaks and `cudaErrorMemoryAllocation` -- however these never caused the script to crash but I'm wondering if they could alter results?

[memcheck.log](https://github.com/user-attachments/files/19320272/memcheck.log)

**Comment by albanD** [comment]
@Apex95 thanks for the detailed search ! It does look suspicious that only one value is wrong.

@eqy could someone on your end double check that we don't do bad 32-bit indexing for layernorm on cuda?

**Comment by mspinaci** [comment]
Hello, I'm adding a minimal reproducible example because this is also affecting me:

```
import torch

device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')

with torch.no_grad():
    x = torch.randn((int(6e6), 768), dtype=torch.float32, device=device)
    layer_norm = torch.nn.LayerNorm(768, eps=1e-7).to(device=device)
    y = layer_norm(x)

print(y[:, 0])
print(torch.where((y == 0).all(dim=1)))
```

This prints
```
tensor([-1.0910, -0.7663, -0.8857,  ...,  0.0000,  0.0000,  0.0000],
       device='cuda:0')
(tensor([5592406, 5592407, 5592408,  ..., 5999997, 5999998, 5999999],
       device='cuda:0'),)
```
i.e. results become flat 0.0 from index `5592406` on. This definitely looks like a uint32 overflow, because:
- 5592406 * 768 = 2**32 + 512
- If I change the vector dim to `1024 = 2**10`, the issue starts happening exactly at index `4194304 = 2**22`
- Running the same code above on CPU works as expected.
